plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2021-2023 Jason Gerecke <jason.gerecke@wacom.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-wac-common.h"
+#include "fu-wac-device.h"
+#include "fu-wac-module-bluetooth-id9.h"
+#include "fu-wac-struct.h"
+
+struct _FuWacModuleBluetoothId9 {
+	FuWacModule parent_instance;
+};
+
+G_DEFINE_TYPE(FuWacModuleBluetoothId9, fu_wac_module_bluetooth_id9, FU_TYPE_WAC_MODULE)
+
+#define FU_WAC_MODULE_BLUETOOTH_ID9_PAYLOAD_SZ	   256
+#define FU_WAC_MODULE_BLUETOOTH_ID9_START_NORMAL   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL	   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE  0xFE
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_RAM	   0x02
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_BEGIN   0x03
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_DATA	   0x04
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL 0xEDB88320
+
+typedef struct __attribute__((packed)) {
+	guint16 unknown1;
+	guint32 unknown2;
+	guint16 size;
+} FuWacModuleBluetoothId9UnknownCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 spi_cmd;
+	guint32 start_addr;
+	guint16 size;
+	FuWacModuleBluetoothId9UnknownCmd data;
+
+} FuWacModuleBluetoothId9SpiCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 command;
+	guint16 size;
+	guint32 crc;
+	FuWacModuleBluetoothId9SpiCmd data;
+} FuWacModuleBluetoothId9LoaderCmd;
+
+static GBytes *
+fu_wac_module_bluetooth_id9_get_startcmd(const guint8 *data, gsize sz, gboolean full_erase)
+{
+	FuWacModuleBluetoothId9LoaderCmd cmd = {
+	    .command = full_erase ? FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE
+				  : FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL,
+	    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data) + sz)),
+	    .crc = 0,
+	    .data =
+		{
+		    .spi_cmd = 0x91,
+		    .start_addr = 0,
+		    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data.data) + sz)),
+		    .data =
+			{
+			    .unknown1 = GUINT16_TO_BE((guint16)0x7050),
+			    .unknown2 = 0,
+			    .size = GUINT16_TO_BE((guint16)sz),
+			},
+		},
+	};
```

Comment:
I think this would be much better using rustgen -- e.g. see https://github.com/fwupd/fwupd/blob/main/docs/tutorial.md#the-rustgen-helper

Reply:
I'm starting work on the conversion, and have noticed that rustgen doesn't seem to support nested structures. That is, if I try to define `struct FuWacModuleBluetoothId9SpiCmd` as containing a `FuWacModuleBluetoothId9UnknownCmd` field, it fails to compile with an "invalid type" error. I can flatten the structure, but makes the "size" calculation a bit more opaque.

Reply:
I agree that's not awesome. Do the workaround for now and I'll work on nested structs tomorrow.

Reply:
@jigpu if you merge in the commit from the wip/hughsie/rustgen-nested-types branch does that work?

Reply:
@hughsie Yes, I'm able to get the nested types working with that commit merged in. Do you want me to rebase my code on top of that branch?

Reply:
yup, please